### ISSUE TITLE
[Question] in A2C，why use “evaluate_actions” to duplicatedly compute log_probs and values

### ISSUE THREAD

**Issue by ABC67876** [DLR-RM/stable-baselines3#1339]
### ❓ Question

In **/stable_baselines3/a2c/a2c.py**, after collecting rollout buffer in function "learn", we get the actions and advantages from rollout buffer, then compute the loss.
**However, why use the "evalutate_actions" to duplicatedly compute values and log_probs?** Can we just get these values from rollout buffer?
```
def train(self) -> None:
        # This will only loop once (get all data in one go)
        for rollout_data in self.rollout_buffer.get(batch_size=None):
            actions = rollout_data.actions
            # ......
            values, log_prob, entropy = self.policy.evaluate_actions(rollout_data.observations, actions)

            # Policy gradient loss
            policy_loss = -(advantages * log_prob).mean()

            # Value loss using the TD(gae_lambda) target
            value_loss = F.mse_loss(rollout_data.returns, values)
            # ......
```

**Question 1**: According to my understanding, A2C is **on-policy and synchronized**, so the policy won't update during the buffer collecting, then this step is unnecessary. And this may causes problems if the network includes random factors (e.g. dropout).

I've carefully read related documentation, except for the part about Vectorized Environments[¶](https://stable-baselines3.readthedocs.io/en/master/guide/vec_envs.html#vectorized-environments)    

I've also read the implementation of **rolloutbuffer**, which returns a RolloutBufferSamples:
```
class RolloutBufferSamples(NamedTuple):
    observations: th.Tensor
    actions: th.Tensor
    old_values: th.Tensor
    old_log_prob: th.Tensor
    advantages: th.Tensor
    returns: th.Tensor
```
**Question 2**: I can't understand why the terms are called "**old**" values and logprob, in a buffer used for on-policy algorithms. 

Maybe these 2 problems above are related.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by qgallouedec** [comment]
In on-policy algorithms, rollout data is split into mini-batches.

https://github.com/DLR-RM/stable-baselines3/blob/12e9917c24dc23d7de7694a924f017c6a8e9a6ce/stable_baselines3/a2c/a2c.py#L133

So your remarks are true, but only for the first minibatch. After each minibatch, the policy changes, so you need to recompute the value and log-prob.

**Comment by ABC67876** [comment]
> In on-policy algorithms, rollout data is split into mini-batches.


Thank you for your reply! **That really makes sense**, and also explains why the data is **shuffled** in "get":
https://github.com/DLR-RM/stable-baselines3/blob/12e9917c24dc23d7de7694a924f017c6a8e9a6ce/stable_baselines3/common/buffers.py#L448

Still, I wonder **is mini-batch necessary?**  I've read some tutorials about RL, but mini-batch is not mentioned. 

This will cause huge **overhead** (almost do an extra "forward"). If I discard this step, the whole episode can be considered as a batch. Is it worth a try?

**Comment by qgallouedec** [comment]
After reading once again, I realize that data are split into minibatches only in PPO. In A2C, the entire episode in returned (`batch_size=None`). Consequently, I also wonder why we recompute those values. @araffin any insight on this?

**Comment by araffin** [comment]
Hello,

yes, I'm aware of that. In fact, there use to be a `# TODO: avoid second computation of everything because of the gradient` in `a2c.py`.

The main reason we recompute them is because we need the gradient information.
As the replay buffer uses numpy as backend, the gradient information is lost when storing the log prob/values, so even though we are not doing any mini-batch, we need to recompute those quantities.

Historically, the replay buffer backend was in PyTorch, but then as most values were numpy arrays (coming from gym) and to keep all gradient computation in one place, I decided to switch to numpy (and even with PyTorch as backend, the gradient will be used only once and for A2C only).

Finally, the forward pass in A2C is usually fast, the bottleneck should be in the backward pass. So, it does slow down a bit the code but in most cases (especially when not using CNN), it should not impact too much the speed of A2C.

> Still, I wonder is mini-batch necessary? I've read some tutorials about RL, but mini-batch is not mentioned.

You may have a look at https://iclr-blog-track.github.io/2022/03/25/ppo-implementation-details/, you can also experiment by increasing the mini-batch size until it is a full batch.

**Comment by ABC67876** [comment]
> The main reason we recompute them is because we need the gradient information.

I‘ve never noticed that. Thank you for your explaination.

The blog is very detailed and seems helpful to me.

As my question has been solved, I will close this issue soon. I appreciate all the kind help above.
